Added kubernetes deployment options

loewenstein opened the pull request.

cppforlife commented on kubernetes/bosh-cloud-config.yml:
```diff
@@ -0,0 +1,17 @@
+        ports:
+          - name: nats
+            port: 4222
+            nodePort: ((port-nats))
```
general convention within this project: all variables should have underscores instead of dashes.

cppforlife commented on kubernetes/cloud-config.yml:
```diff
@@ -0,0 +1,22 @@
+azs:
+- name: z1
```
let's add z1, z2, z3. see warden for example.

cppforlife commented on kubernetes/cloud-config.yml:
```diff
@@ -0,0 +1,22 @@
+- name: default
+  type: dynamic
+  subnets:
+  - azs: [z1]
```
how is dns configured? defaults to something?

loewenstein replied:
What are you missing specifically? 
/etc/resolve.conf looks like the following on Pods/Containers created by the CPI:
```
nameserver 10.0.0.10
search default.svc.cluster.local svc.cluster.local cluster.local
options ndots:5
```

cppforlife replied:
i guess if we say that all dns is configured by kube then we are done.

cppforlife commented on kubernetes/cloud-config.yml:
```diff
@@ -0,0 +1,22 @@
+  - azs: [z1]
+
+compilation:
+  workers: 3
```
crank it up to 5 or 6.

loewenstein replied:
That was too much for my minikube installation. I can try out how many I could spin up in parallel. (GHC is quite a demanding compiler :) )

cppforlife commented on kubernetes/cpi-inner.yml:
```diff
@@ -0,0 +1,25 @@
+---
```
dont believe this file is necessary given that https://github.com/cloudfoundry/bosh-deployment/blob/master/bosh-dev.yml exists.

loewenstein replied:
Not sure I understand. How would you remove cpi url, make cpi version latest, remove static IPs, etc.?

cppforlife replied:
i guess static_ips aren removed.

cppforlife commented on kubernetes/bosh-cloud-config.yml:
```diff
@@ -0,0 +1,17 @@
+---
```
can we hard code some of these host ports for ease of use?

loewenstein replied:
These ports need to be unique for the whole cluster. And the range is even configurable. We might force people to edit bosh-deployment by hard coding them. I was thinking about adding a minikube ops file. That could actually hard-code all the ports for ease of first usage. What do you think?

cppforlife replied:
sure, i think we want to optimize for 95% case. rest of cases could be done by having ppl write their own custom ops file.

cppforlife commented on kubernetes/cpi.yml:
```diff
@@ -0,0 +1,82 @@
+  path: /instance_groups/name=bosh/properties/kubernetes?
+  value: &kubernetes
+    apiEndpoint: https://((host_ip)):8443
+    namespace: ((namespace))
```
should probably be `kube_namespace`

cppforlife commented on kubernetes/cpi.yml:
```diff
@@ -0,0 +1,82 @@
+  value: &kubernetes
+    apiEndpoint: https://((host_ip)):8443
+    namespace: ((namespace))
+    credentials:
```
convention for cert properties is:

- private_key
- certificate
- ca (dont need this one in this case)

that way can just use `credentials: ((client_cert))`

loewenstein replied:
Isn't it rather unlikely that anyone has got his kubernetes creds in a bosh specified format?

In the README I pass the with two --var-file options
```
...
--var-file=client-cert=$(kubectl config view -o json | jq -r '.users[] | select(.name=="minikube") | .user."client-certificate"') \
--var-file=client-key=$(kubectl config view -o json | jq -r '.users[] | select(.name=="minikube") | .user."client-key"') \
...
```

cppforlife replied:
i think may be then this is blocked on ability to specify var pieces: eg 

```
--var-file kube_client_creds.certificate=...
--var-file kube_client_creds.private_key=...
```

cppforlife commented on kubernetes/cpi.yml:
```diff
@@ -0,0 +1,82 @@
+- type: replace
+  path: /instance_groups/name=bosh/properties/kubernetes?
+  value: &kubernetes
+    apiEndpoint: https://((host_ip)):8443
```
host_ip -> kube_host_ip

loewenstein commented:
- I have renamed variables to use underscore in place of dash
- I have extracted custom port config to separate ops files
- I have added AZs and increased number of compilation workers
- I still need an `cpi-inner.yml` ops file
  - for one I need to add a vm_extension and reconfigure blobstore port
  - in addition I wanted to upload releases and stemcell and use version `latest`

cppforlife commented on kubernetes/cpi.yml:
```diff
@@ -0,0 +1,95 @@
+      namespace: ((kube_namespace))
+      credentials:
+        certificate: ((client_cert))
+        private-key: ((client_key))
```
im not too hot on casing differences. there are dashes, underscores and camel case that's involved. let's convert everything to underscores.
